src/main/java/org/mariadb/jdbc/internal/failover/AbstractMastersListener.java

```diff
@@ -352,7 +416,28 @@ public FailLoop(Listener listener) {
             this.listener = listener;
         }
 
-        public void run() {
+        public void scheduleSelf(ScheduledExecutorService scheduler, boolean now) {
+            scheduledFuture = scheduler.scheduleWithFixedDelay(this, now ? 0 : 250, 250, 
+                                                               TimeUnit.MILLISECONDS);
+        }
+
+        // while scheduledFuture may be set from another thread
+        // it is expected only one thread will unschedule
+        @Override
+        public void unscheduleTask() {
+            while (true) {
+                if (scheduledFuture == null) {
+                    // not scheduled yet, should be a rare condition, just loop
+                    Thread.yield();
+                } else {
+                    scheduledFuture.cancel(false);
+                    scheduledFuture = null;
```

Comment:
a "return" after this seems good : If failLoop reconnect the failed connection, the stopFailover() (actual line 451) will loop inside this while, and never freed the thread.